scriptworker/artifacts.py

```diff
@@ -331,18 +331,35 @@ def get_upstream_artifacts_full_paths_per_task_id(context):
         scriptworker.exceptions.ScriptWorkerTaskException: when an artifact doesn't exist.
 
     """
+    upstream_artifacts = context.task['payload']['upstreamArtifacts']
     task_ids_and_relative_paths = [
         (artifact_definition['taskId'], artifact_definition['paths'])
-        for artifact_definition in context.task['payload']['upstreamArtifacts']
+        for artifact_definition in upstream_artifacts
     ]
 
-    return {
-        task_id: [
-            get_and_check_single_upstream_artifact_full_path(context, task_id, path)
-            for path in paths
-        ]
-        for task_id, paths in task_ids_and_relative_paths
-    }
+    optional_artifacts_per_task_id = get_optional_artifacts_per_task_id(upstream_artifacts)
+
+    upstream_artifacts_full_paths_per_task_id = {}
+    failed_paths_per_task_id = {}
+    for task_id, paths in task_ids_and_relative_paths:
+        for path in paths:
+            try:
+                path_to_add = get_and_check_single_upstream_artifact_full_path(context, task_id, path)
+                add_enumerable_item_to_dict(
+                    dict_=upstream_artifacts_full_paths_per_task_id,
+                    key=task_id, item=path_to_add
+                )
+            except ScriptWorkerTaskException:
+                if path in optional_artifacts_per_task_id.get(task_id, []):
+                    log.warn('Optional artifact "{}" of task "{}" not found'.format(path, task_id))
+                    add_enumerable_item_to_dict(
+                        dict_=failed_paths_per_task_id,
+                        key=task_id, item=path
+                    )
+                else:
+                    raise
+
+    return upstream_artifacts_full_paths_per_task_id, failed_paths_per_task_id
```

Comment:
You're changing the return type of a non-private function. I think pushapkscript is the only consumer, so we're probably ok here.

1. this means this will be a major version bump
2. Please update the docstring.

Reply:
Ooops, I thought I did. Thanks for spotting it. It's now fixed.